### PR TITLE
Use native and/or fake `harden` for performance

### LINKED ISSUE #7275: Adopt XS native `harden` for vats
## What is the Problem Being Solved?
Vaults scenario testing has shown that native XS harden is roughly 20% faster than SES-shim’s harden on XS:

| worker | vaultsOpened | duration (ms) | avgPerVault (ms) | 
| --- | --- | --- | --- |
| local | 100 | 32147.71866607666 | 321.4771866607666 |
| xs-shim | 100 | 90076.88854074478 | 900.7688854074478 |
| xs-native | 100 | 70358.89608383179 | 703.5889608383179 |

generated via `git checkout ari/perf_vaults packages/vats/test/bootstrapTests/test-vaults-perf.js` and ran in different configurations

## Description of the Design
Experiment in [mfig-native-harden](https://github.com/Agoric/agoric-sdk/tree/mfig-native-harden) is the branch
- Change is in SES-shim lockdown under XS. If `globalThis.harden` exists before `lockdown` then use it. Endo tests, swingset tests.

Notes from discussion:
- can we run moddable lockdown after SES shim lockdown without problems?
- Date.now replacement

## Security Considerations
Will need to ensure this implementation of `harden` meets spec to maintain immutability

## Scaling Considerations
Need to re-run testing to ensure CPU impact is as expected & memory usage isn't off

### LINKED ISSUE #7367: use fake harden() in chain's kernel process
## What is the Problem Being Solved?

We think that `harden()` is consuming non-trivial CPU time, and causing a measurable performance impact. In our `xsnap` workers, we'll be able to improve this considerably by using the native XS `harden` feature (#7275). But in the kernel process, which runs under Node.js/V8, that's not an option.

Our kernel process is not running untrusted code. Good ocap discipline encourages us to harden all API surfaces, but we don't strictly need it to defend against attacking code. @erights suggested that it'd be safe to replace the kernel-side `harden()` calls with a NOP, for the sake of performance improvement. He's made changes to the handful of places that asserted `Object.isFrozen`/etc, so at this point, nothing on the kernel side should notice that `harden()` is not actually doing anything. Endo has a `lockdown()` option to replace `harden()` with a NOP, or maybe it looks at `globalThis.process.env`.

## Description of the Design

As a performance -vs- nominal safety tradeoff, this decision should be made by the host application author (not swingset internally). For our chain, that means in `packages/cosmic-swingset/src/entrypoint.js`, where it imports `@endo/init`:

https://github.com/Agoric/agoric-sdk/blob/2b31d207b10b99444da630670d3b7dd2e6d9356f/packages/cosmic-swingset/src/entrypoint.js#L7

@michaelfig suggested creating an alternate Endo initialization package, similar to our `@endo/init/debug.js`, to disable harden. Once that exists, this `entrypoint.js` would instead do something like:

```js
import '@endo/init/no-harden.js`;
```

We might also consider allowing `globalThis.process.env` to control some aspect of this.

## Security Considerations

We must make sure that whatever knob we introduce to disable `harden` does not reduce the security of mutually-suspicious-code environments. If we were using Node.js subprocess vat workers (instead of xsnap workers), we'd need to be worried about environment variables being inherited by child processes. Xsnap does not populate `globalThis.process.env` (the whole `process` space is a Node.js thing).

## Scaling Considerations

This should speed up the kernel process. We should have some benchmarks to demonstrate exactly how much of a speedup we're buying.

## Test Plan

Having a `console.log` in the Endo code that announces the disabling of `harden()` would be sufficient: just check that this message appears when the chain is launched.

### PR DESCRIPTION
**Pull request by michaelfig**
closes: #7275, closes: #7367
refs: agoric-labs/xsnap-pub#37 agoric-labs/moddable#8 endojs/endo#1556

## Description

Optimise the on-chain use of `harden`

- Use native `globalThis.harden` under XS (agoric-labs/xsnap-pub#37 and agoric-labs/moddable#8)
- Use `@endo/init/unsafe-fast.js` for Node.js cosmic-swingset endojs/endo#1528 endojs/endo#1552

Please note the Git hash updates in this PR, and see that the submodule diffs below are as expected.

### Security Considerations

Some softer interface boundaries within the SwingSet kernel (running in Node.js) since `harden` is faked there for performance.  We consider this to be an acceptable risk, because the kernel is carefully written, and does not evaluate any externally-supplied code except within a vat worker (which still uses the full, safe `harden` implementation).

### Scaling Considerations

Reduces runtime costs of `harden()` under `xsnap-worker` and within the SwingSet kernel launched by `agd start`.

### CODE REVIEW

**Comment by mhofman** [comment]
I'll waive the requirement to bump the meter version given we will soon update XS versions anyway.

**Review by erights — COMMENTED** [review]
Swingset needs to continue to work with normal safe harden as well, for general correctness and sanity. Can we have a way to run the Swingset tests with safe harden too? Should this happen in CI?

**Comment by michaelfig** [comment]
> Swingset needs to continue to work with normal safe harden as well, for general correctness and sanity. Can we have a way to run the Swingset tests with safe harden too? Should this happen in CI?

The only change to use unsafe harden is in the entry points in the `cosmic-swingset` package (the code used when running `agd`).  Everywhere else, including all CI, normal safe harden is used.

File: packages/SwingSet/test/snapshots/test-xsnap-store.js.md [hunk 1]
```diff
@@ -10,25 +10,25 @@ Generated by [AVA](https://avajs.dev).
...
 
 > after SES boot - sensitive to SES-shim, XS, and supervisor
 
     {
       compressSeconds: 0,
-      hash: 'fd8d35d908d4d5892fb049edec5867ebd737e21be1bf8a4ae2bd810f81260061',
+      hash: 'caf69d62160347624ec1680d819db5152fd9e766e3f8e3f87c5a1e95ee2c1cbf',
       rawSaveSeconds: 0,
-      uncompressedSize: 801387,
+      uncompressedSize: 724579,
```
Comment (mhofman) [inline packages/SwingSet/test/snapshots/test-xsnap-store.js.md:24]: Impressive how we reduced the heap snapshot size!

**Review by mhofman — APPROVED** [review]
Neat, clean change!

File: packages/SwingSet/test/snapshots/test-xsnap-store.js.md (hunk 1 above)
Comment (michaelfig) [reply packages/SwingSet/test/snapshots/test-xsnap-store.js.md:24]: Yeah, removing the WeakSet of hardened values is probably what caused this shrinkage.

**Review by warner — APPROVED** [review]
yep, looks good, awesome that this works!

**Review by erights — APPROVED** [review]
LGTM Full speed ahead!

**Pull request merged**